File: app/src/main/java/com/danilomendes/figurines/ui/base/AbstractPresenter.java
package com.danilomendes.figurines.ui.base;

import android.support.annotation.NonNull;

import com.danilomendes.figurines.utils.L;

import io.reactivex.disposables.CompositeDisposable;

/**
 * Created by danilo on 18-10-2017.
 */
public abstract class AbstractPresenter<T> {
    protected T mView;
    protected CompositeDisposable mCompositeDisposable;

    public final void attachView(@NonNull T view) {
        mView = view;
        mCompositeDisposable = new CompositeDisposable();
    }

    public final void detachView() {
        mView = null;
        if (mCompositeDisposable != null && !mCompositeDisposable.isDisposed()) {
            L.log("Dispose disposables.");
            mCompositeDisposable.dispose();
        }
    }

    protected final boolean isViewAttached() {
        return mView != null;
    }
}

File: app/src/main/java/com/danilomendes/figurines/ui/base/AbstractPresenter.kt
package com.danilomendes.figurines.ui.base

import com.danilomendes.figurines.util.L
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by danilo on 08-12-2017.
 */
abstract class AbstractPresenter<T> {

    protected var view: T? = null

    protected val compositeDisposable: CompositeDisposable by lazy { CompositeDisposable() }

    fun attachView(view: T) {
        L.log("Attaching view.")
        this.view = view
    }

    fun detachView() {
        this.view = null
        if (!compositeDisposable.isDisposed) {
            compositeDisposable.dispose()
        }
    }

    fun isViewAttached(): Boolean {
        return this.view != null
    }
}